README.md
# [![Send](./assets/icon-64x64.png)](https://gitlab.com/timvisee/send/) Send

A fork of Mozilla's [Firefox-Send](https://github.com/mozilla/send). 
<br>
Simple, private AES-128-GCM end-to-end encrypted file sharing service.

### €⁠20 on [Hetzner Cloud](https://hetzner.cloud/?ref=eLtKhFK70n4h)

### Automatic Installs
```
https://github.com/WhateverItWorks/Watchtower
```

## Deploy with Docker
Ensure [Docker Engine](https://docs.docker.com/engine/install/) and [Docker Compose](https://docs.docker.com/compose/install/) are installed before beginning.

Run send in production with docker-compose:
```
apt install git
git clone https://github.com/timvisee/send.git
cd send
nano docker-compose.yml
chmod 777 uploads
docker-compose pull
docker-compose up -d
```
Should be running at: http://localhost:1234

Run send in developer with docker-compose:
```
apt install git
git clone https://github.com/timvisee/send.git
cd send
nano docker-compose.yml
chmod 777 uploads
docker-compose up -d --build
```
Should be running at: http://localhost:1234


## Environment Variables

All the available config options and their defaults can be found here: https://github.com/timvisee/send/blob/master/server/config.js

Config options should be set as unquoted environment variables. Boolean options should be `true`/`false`, time/duration should be integers (seconds), and filesize values should be integers (bytes).

Config options expecting array values (e.g. `EXPIRE_TIMES_SECONDS`, `DOWNLOAD_COUNTS`) should be in unquoted CSV format. UI dropdowns will default to the first value in the CSV, e.g. `DOWNLOAD_COUNTS=5,1,10,100` will show four dropdown options, with `5` selected by the default.

#### Server Configuration

| Name     | Description |
|------------------|-------------|
| `BASE_URL`       | The HTTPS URL where traffic will be served (e.g. `https://send.firefox.com`)
| `DETECT_BASE_URL` | Autodetect the base URL using browser if `BASE_URL` is unset (defaults to `false`)
| `PORT`           | Port the server will listen on (defaults to `1443`)
| `NODE_ENV`       | Run in `development` mode (unsafe) or `production` mode (the default)
| `SEND_FOOTER_DMCA_URL` | A URL to a contact page for DMCA requests (empty / not shown by default)
| `SENTRY_CLIENT`, `SENTRY_DSN`  | Sentry Client ID and DSN for error tracking (optional, disabled by default)

*Note: more options can be found here: https://github.com/timvisee/send/blob/master/server/config.js*

#### Upload and Download Limits

Configure the limits for uploads and downloads. Long expiration times are risky on public servers as people may use you as free hosting for copyrighted content or malware (which is why Mozilla shut down their `send` service). It's advised to only expose your service on a LAN/intranet, password protect it with a proxy/gateway, or make sure to set `SEND_FOOTER_DMCA_URL` above so you can respond to takedown requests.

| Name    | Description |
|------------------|-------------|
| `MAX_FILE_SIZE` | Maximum upload file size in bytes (defaults to `2147483648` aka 2GB)
| `MAX_FILES_PER_ARCHIVE` | Maximum number of files per archive (defaults to `64`)
| `MAX_EXPIRE_SECONDS` | Maximum upload expiry time in seconds (defaults to `604800` aka 7 days)
| `MAX_DOWNLOADS` | Maximum number of downloads (defaults to `100`)
| `DOWNLOAD_COUNTS` | Download limit options to show in UI dropdown, e.g. `10,1,2,5,10,15,25,50,100,1000`
| `EXPIRE_TIMES_SECONDS` | Expire time options to show in UI dropdown, e.g. `3600,86400,604800,2592000,31536000`
| `DEFAULT_DOWNLOADS` | Default download limit in UI (defaults to `1`)
| `DEFAULT_EXPIRE_SECONDS` | Default expire time in UI (defaults to `86400`)

*Note: more options can be found here: https://github.com/timvisee/send/blob/master/server/config.js*

#### Storage Backend Options

Pick how you want to store uploaded files and set these config options accordingly:

- Local filesystem (the default): set `FILE_DIR` to the local path used inside the container for storage (or leave the default)
- S3-compatible object store: set `S3_BUCKET`, `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` (and `S3_ENDPOINT` if using something other than AWS)
- Google Cloud Storage: set `GCS_BUCKET` to the name of a GCS bucket (auth should be set up using [Application Default Credentials](https://cloud.google.com/docs/authentication/production#auth-cloud-implicit-nodejs))

Redis is used as the metadata database for the backend and is required no matter which storage method you use.

| Name  | Description |
|------------------|-------------|
| `REDIS_HOST`, `REDIS_PORT`, `REDIS_USER`, `REDIS_PASSWORD`, `REDIS_DB` | Host name, port, and pass of the Redis server (defaults to `localhost`, `6379`, and no password)
| `FILE_DIR`       | Directory for storage inside the Docker container (defaults to `/uploads`)
| `S3_BUCKET`  | The S3 bucket name to use (only set if using S3 for storage)
| `S3_ENDPOINT` | An optional custom endpoint to use for S3 (defaults to AWS)
| `S3_USE_PATH_STYLE_ENDPOINT`| Whether to force [path style URLs](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html#s3ForcePathStyle-property) for S3 objects (defaults to `false`)
| `AWS_ACCESS_KEY_ID` | S3 access key ID (only set if using S3 for storage)
| `AWS_SECRET_ACCESS_KEY` | S3 secret access key ID (only set if using S3 for storage)
| `GCS_BUCKET` | Google Cloud Storage bucket (only set if using GCP for storage)

*Note: more options can be found here: https://github.com/timvisee/send/blob/master/server/config.js*

## Branding

To change the look the colors aswell as some graphics can be changed via environment variables.  
See the table below for the variables and their default values.

| Name | Default | Description |
|---|---|---|
| UI_COLOR_PRIMARY | #0a84ff | The primary color |
| UI_COLOR_ACCENT | #003eaa | The accent color (eg. for hover-effects) |
| UI_CUSTOM_ASSETS_ANDROID_CHROME_192PX | | A custom icon for Android (192x192px) |
| UI_CUSTOM_ASSETS_ANDROID_CHROME_512PX | | A custom icon for Android (512x512px) |
| UI_CUSTOM_ASSETS_APPLE_TOUCH_ICON | | A custom icon for Apple |
| UI_CUSTOM_ASSETS_FAVICON_16PX | | A custom favicon (16x16px) |
| UI_CUSTOM_ASSETS_FAVICON_32PX | | A custom favicon (32x32px) |
| UI_CUSTOM_ASSETS_ICON | | A custom icon (Logo on the top-left of the UI) |
| UI_CUSTOM_ASSETS_SAFARI_PINNED_TAB | | A custom icon for Safari |
| UI_CUSTOM_ASSETS_FACEBOOK | | A custom header image for Facebook |
| UI_CUSTOM_ASSETS_TWITTER | | A custom header image for Twitter |
| UI_CUSTOM_ASSETS_WORDMARK | | A custom wordmark (Text next to the logo) |
| UI_CUSTOM_CSS | | Allows you to define a custom CSS file for custom styling |
| CUSTOM_FOOTER_TEXT | | Allows you to define a custom footer |
| CUSTOM_FOOTER_URL | | Allows you to define a custom URL in your footer |

Side note: If you define a custom URL and a custom footer, only the footer text will display, but will be hyperlinked to the URL.

## Security Audits:

- [Internet.nl](https://internet.nl/site/send.whateveritworks.org/2060148/)
- [HSTS Preload](https://hstspreload.org/)
- [SSL Labs](https://www.ssllabs.com/ssltest/analyze.html?d=send.whateveritworks.org)
- [Security Headers](https://securityheaders.com/?q=send.whateveritworks.org&hide=on&followRedirects=on)
- [pagespeed](https://pagespeed.web.dev/)
- [webbkoll](https://webbkoll.dataskydd.net/en)
- [ImmuniWeb](https://www.immuniweb.com/ssl/send.whateveritworks.org/uLlrAeMb/)
- [Hardenize](https://www.hardenize.com/report/send.whateveritworks.org/1686343966)
- [Mozilla.org](https://observatory.mozilla.org/)
- [report-uri.com](https://report-uri.com/home/tools)
- [check-your-website.server-daten.de](https://check-your-website.server-daten.de/?q=search.whateveritworks.org)
- [csp-evaluator.withgoogle.com](https://csp-evaluator.withgoogle.com/)
- [OpenWPM](https://github.com/openwpm/OpenWPM)
- [privacyscore.org](https://privacyscore.org/site/215029/)

## Instances

- [Instances](#instances)
- [Live Status](#live-status)
- [How to use](#how-to-use-cli)
- [Host your own instance](#deploy-with-docker)
- [Security Audits](#security-audits)
- [Submit changes](#submit-changes)
- [DMCA Takedown Process Request](#dmca-takedown-process-request)

This page does not give any promises or warranties with regard to instance
security and reliability.

Instance URL | Size<br>limit | Time<br>limit | DL<br>limit | Links/<br>Notes | Country | Version | Uptime<br>(90 days)
--- | ---: | ---: | ---: | --- | ---: | --- | ---
https://send.vis.ee | 2.5GiB | 3 days | 10 | [maintainer](https://github.com/timvisee), [contact](https://timvisee.com/contact) | Netherlands 🇳🇱 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.vis.ee/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230691-8f60854620eb9d40dae7461e)
https://send.zcyph.cc | 10GiB | 7 days | 100 | [maintainer](https://github.com/zcyph), [contact](mailto:send@zcyph.cc) | Germany 🇩🇪 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.zcyph.cc/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230695-ef83cfab5f4970c4487ad484)
https://send.ephemeral.land | 8GiB | 28 days | 1,000 | | Germany 🇩🇪 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.ephemeral.land/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230698-d0fc0e9c893bdb81295c3ae2)
https://send.mni.li | 8GiB | 7 days | 25 | [contact](https://cryptpad.fr/form/#/2/form/view/gj2mDNekg5gf+AKPkTqLGY9W2Fa2rjceLFISeeLZa3Y/) | Germany 🇩🇪 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.mni.li/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230704-256d7241d3fc3712ed74671c)
https://send.monks.tools | 5GiB | 7 days | 50 | | United States 🇺🇸 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.monks.tools/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230706-152180f0f00c3516167167fd)
https://send.boblorange.net | 2.5GiB | 7 days | 100 | | Portugal 🇵🇹 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.boblorange.net/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230709-6647754935bde8c9e48f74b0)
https://send.aurorabilisim.com | 2.5GiB | 7 days | 100 | [contact](https://www.aurorabilisim.com/iletisim/) | Turkey 🇹🇷 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.aurorabilisim.com/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230712-50b203eb6aba33e7ed9f35a4)
https://send.artemislena.eu | 2.5GiB | 7 days | 100 | [contact](https://artemislena.eu/contact.html) | Germany 🇩🇪 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.artemislena.eu/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230717-7d15b7ccd3aa5630bba41a8a)
https://send.datahoarder.dev | 1GiB | 1 day | 5 | [maintainer](https://github.com/whalehub), [contact](mailto:admin@datahoarder.dev) | Luxembourg 🇱🇺 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.datahoarder.dev/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230721-b9f403d93360e8d40a50128d)
https://fileupload.ggc-project.de | 2.5GiB | 7 days | 100 | | Germany 🇩🇪 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://fileupload.ggc-project.de/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230727-a652ecfbd4604f532e0ff48c)
https://drop.chapril.org | 1GiB | 5 days | 100 | [contact](https://www.chapril.org/contact.html) | Germany 🇩🇪 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://drop.chapril.org/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230729-fd070e4aa8a6601035d9425a)
https://send.jeugdhulp.be | 50MiB | 10 days | 25 | [contact](https://www.jeugdhulp.be/contact) | France 🇫🇷 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.jeugdhulp.be/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230732-350359a789fe5d02bdb5ad6d)
https://files.psu.ru | 16GiB | 7 days | 500 | no password | Russia 🇷🇺 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://files.psu.ru/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230735-2d39e5d423d54366178406da)
https://send.portailpro.net | 10GiB | 30 days | 100 | | France 🇫🇷 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.portailpro.net/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230737-577e6bfd57f73d339fd6554f)
https://transfer.acted.org | 5GiB | 14 days | 3,000 | | France 🇫🇷 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://transfer.acted.org/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230739-463fdbc2d9c115069d8db1f1)
https://send.datenpost.app | 30GiB | 7 days | 3 | [contact](mailto:info@webality.de) | Germany 🇩🇪 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.datenpost.app/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230691-8f60854620eb9d40dae7461e)
https://send.angelic.icu | 2.5GiB | 7 days | 50 | [contact](mailto:me@angelic.icu) | Romania 🇷🇴 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.angelic.icu/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794230747-a87dcbdff5b01eb9c5f92b6c)
https://s.opnxng.com | 2.5GiB | 7 days | 25 | [contact](https://about.opnxng.com/) | Singapore 🇸🇬 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://s.opnxng.com/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794398378-baf9f42c4a7e416bc51f5ba0)
https://send.whateveritworks.org | 10GiB | 7 days | 100 | [contact](https://www.whateveritworks.org/email) | Germany 🇩🇪 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.whateveritworks.org/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794407638-626e4c3452c6933ad106a402)
https://send.cyberjake.xyz | 10GiB | 30 days | 1000 | [contact](mailto:connect@cyberjake.xyz) | United States 🇺🇸 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.cyberjake.xyz/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794455795-2d6b721c8e5dfcc8dcdae877)
https://send.kokomo.cloud | 2.5GiB | 7 days | 100 | [maintainer](https://github.com/kokomo123), [contact](mailto:admin@kokomo.cloud) | United States 🇺🇸 | ![version](https://img.shields.io/badge/dynamic/json?label=version&query=version&url=https://send.kokomo.cloud/__version__) | ![Uptime (90 days)](https://img.shields.io/uptimerobot/ratio/90/m794630156-5be85f191fc02e133c49732f)

Users can view the historic reliability of the Send instances on [this status page](https://stats.uptimerobot.com/5917xHMX01). Click each instance on that page for more details.

If you plan to host it publicly, please consider to add it to this list.

## How to Use ```CLI```
To use a specific instance from the command line with [ffsend][ffsend], provide
the `--host URL` flag while uploading.

```
# ffsend upload to custom host
ffsend upload --host https://send.vis.ee/ test.zip
```

## Submit changes

To submit changes to this list, please open a pull request or issue.

[send]: https://github.com/timvisee/send
[ffsend]: https://github.com/timvisee/ffsend

## Live Status

See Uptime Liability Send Instances: [here](https://github.com/tdulcet/send-instances-status)

## DMCA Takedown Process Request

In cases of a DMCA notice, or other abuse yet to be determined, a file has to be removed from the service.

Files can be delisted and made inaccessible by removing their record from Redis.

Send share links contain the `id` of the file, for example `https://send.firefox.com/download/3d9d2bb9a1`

From a host with access to the Redis server run a `DEL` command with the file id.

For example:

```sh
redis-cli DEL 3d9d2bb9a1
```

Other redis-cli parameters like `-h` may also be required. See [redis-cli docs](https://redis.io/topics/rediscli) for more info.


CHANGELOG.md
## Change Log

### v2.5.4 (2023/07/24 19:26 +00:00)
- Renamed LICENSE to LICENSE.md and Changed to AGPL-3.0
- Updated Dependencies
- Optimized code to understand
- Added example nginx.conf
- Harden docker-compose.yml
- Added SECURITY.md Policy
- Updated Nodejs & Optimized Dockerfile
- Optimzed README.md

### v2.5.1 (2018/03/12 19:26 +00:00)
- [#789](https://github.com/mozilla/send/pull/789) Fixed #775 : Made text not-selectable (@RCMainak)

### v2.5.0 (2018/03/08 19:31 +00:00)
- [#782](https://github.com/mozilla/send/pull/782) updated docs (@dannycoates)
- [#781](https://github.com/mozilla/send/pull/781) Don't translate URL-safe chars, b64 is doing it for us (@timvisee)
- [#779](https://github.com/mozilla/send/pull/779) implemented crypto polyfills for ms edge (@dannycoates)

### v2.4.1 (2018/02/28 17:05 +00:00)
- [#777](https://github.com/mozilla/send/pull/777) use a separate circle in the progress svg for indefinite progress (@dannycoates)

### v2.4.0 (2018/02/27 01:55 +00:00)
- [#769](https://github.com/mozilla/send/pull/769) removed unsafe-inline styles via svgo-loader (@dannycoates)
- [#767](https://github.com/mozilla/send/pull/767) added coverage artifact to circleci (@dannycoates)
- [#766](https://github.com/mozilla/send/pull/766) Some frontend unit tests [WIP] (@dannycoates)
- [#761](https://github.com/mozilla/send/pull/761) added maxPasswordLength and passwordError messages (@dannycoates)
- [#764](https://github.com/mozilla/send/pull/764) added indefinite progress mode (@dannycoates)
- [#760](https://github.com/mozilla/send/pull/760) refactored css: phase 1 (@dannycoates)
- [#759](https://github.com/mozilla/send/pull/759) Switch en-US FTL file to new syntax (@flodolo)
- [#758](https://github.com/mozilla/send/pull/758) refactored server (@dannycoates)
- [#757](https://github.com/mozilla/send/pull/757) Update to fluent 0.4.3 (@stasm)

### v2.3.0 (2018/02/01 23:27 +00:00)
- [#536](https://github.com/mozilla/send/pull/536) use redis expire event to delete stored data immediately (@ehuggett)
- [#744](https://github.com/mozilla/send/pull/744) Gradient experiment (@dannycoates)
- [#739](https://github.com/mozilla/send/pull/739) added /api/info/:id route (@dannycoates)
- [#737](https://github.com/mozilla/send/pull/737) big refactor (@dannycoates)
- [#722](https://github.com/mozilla/send/pull/722) Add localization note to 'Time' and 'Downloads' string (@flodolo)
- [#721](https://github.com/mozilla/send/pull/721) show download Limits on page; Fixes #661 (@shikhar-scs)
- [#694](https://github.com/mozilla/send/pull/694) Passwords can now be changed (#687) (@himanish-star)
- [#702](https://github.com/mozilla/send/pull/702) Restricted the banner from showing on unsupported browsers (@himanish-star)
- [#701](https://github.com/mozilla/send/pull/701) improved popup for mobile display; Fixes #699 (@shikhar-scs)
- [#683](https://github.com/mozilla/send/pull/683) API changes to accommodate 3rd party clients (@ehuggett)
- [#698](https://github.com/mozilla/send/pull/698) Popup for delete button attached (@himanish-star)
- [#695](https://github.com/mozilla/send/pull/695) Show Warning, Cancel and Redirect on size > 2GB ; fixes #578 (@shikhar-scs)
- [#684](https://github.com/mozilla/send/pull/684) delete btn popup attached (@himanish-star)
- [#686](https://github.com/mozilla/send/pull/686) Hide password while Typing and after Entering: Fixes #670 (@shikhar-scs)
- [#679](https://github.com/mozilla/send/pull/679) changed font to sans sherif: Solves #676 (@shikhar-scs)
- [#693](https://github.com/mozilla/send/pull/693) README: Fix query link for "good first bugs" (@jspam)
- [#685](https://github.com/mozilla/send/pull/685) checkbox now has a hover effect: fixes #635 (@himanish-star)
- [#668](https://github.com/mozilla/send/pull/668) Add possibility to bind to a specific IP address (@TwizzyDizzy)
- [#682](https://github.com/mozilla/send/pull/682) [Docs] - README.md - minor spelling fixes (@tmm2018)
- [#672](https://github.com/mozilla/send/pull/672) Use EXPIRE_SECONDS to calculate file ttl for static content (@derektamsen)
- [#680](https://github.com/mozilla/send/pull/680) adjusted line height of label : fixes #609 (@himanish-star)

### v2.2.2 (2017/12/19 18:06 +00:00)
- [#667](https://github.com/mozilla/send/pull/667) Make develop the default NODE_ENV (@claudijd)

### v2.2.1 (2017/12/08 18:00 +00:00)
- [#665](https://github.com/mozilla/send/pull/665) stop drag target from flickering when dragging over children (@ericawright)

### v2.2.0 (2017/12/06 23:57 +00:00)
- [#654](https://github.com/mozilla/send/pull/654) Multiple download UI (@dannycoates)
- [#650](https://github.com/mozilla/send/pull/650) #634: overwrite appearance of password submit input (@ovlb)
- [#649](https://github.com/mozilla/send/pull/649) #609 share interface: align text in input and button (@ovlb)

### v2.1.2 (2017/11/16 19:03 +00:00)
- [#645](https://github.com/mozilla/send/pull/645) Remove the leak of the password into the console (@laurentj)

### v2.1.0 (2017/11/15 03:07 +00:00)
- [#641](https://github.com/mozilla/send/pull/641) Added experiment for firefox download promo (@dannycoates)
- [#640](https://github.com/mozilla/send/pull/640) use fluent-langneg for subtag support (@dannycoates)
- [#639](https://github.com/mozilla/send/pull/639) wrap number localization in try/catch (@dannycoates)

### v2.0.0 (2017/11/08 05:31 +00:00)
- [#633](https://github.com/mozilla/send/pull/633) Keyboard navigation/visual feedback regression (@ehuggett)
- [#632](https://github.com/mozilla/send/pull/632) display the 'add password' button only when the input field isn't empty (@dannycoates)
- [#626](https://github.com/mozilla/send/pull/626) Partial fix for #623 (@ehuggett)
- [#624](https://github.com/mozilla/send/pull/624) set a default MIME type in file metadata (@ehuggett)
- [#612](https://github.com/mozilla/send/pull/612) Password UI nits (@dannycoates, @ericawright)

server/config.js
const fs = require('fs');
const convict = require('convict');
const convict_format_with_validator = require('convict-format-with-validator');
const { tmpdir } = require('os');
const path = require('path');
const { randomBytes } = require('crypto');

convict.addFormats(convict_format_with_validator);

convict.addFormat({
  name: 'positive-int-array',
  coerce: ints => {
    // can take: int[] | string[] | string (csv), returns -> int[]
    const ints_arr = Array.isArray(ints) ? ints : ints.trim().split(',');
    return ints_arr.map(int =>
      typeof int === 'number'
        ? int
        : parseInt(int.replace(/['"]+/g, '').trim(), 10)
    );
  },
  validate: ints => {
    // takes: int[], errors if any NaNs, negatives, or floats present
    for (const int of ints) {
      if (typeof int !== 'number' || isNaN(int) || int < 0)
        throw new Error('must be a comma-separated list of positive integers');
    }
  }
});

const conf = convict({
  s3_bucket: {
    format: String,
    default: '',
    env: 'S3_BUCKET'
  },
  s3_endpoint: {
    format: String,
    default: '',
    env: 'S3_ENDPOINT'
  },
  s3_use_path_style_endpoint: {
    format: Boolean,
    default: false,
    env: 'S3_USE_PATH_STYLE_ENDPOINT'
  },
  s3_logging_enabled: {
    format: Boolean,
    default: false,
    env: 'S3_LOGGING_ENABLED'
  },
  gcs_bucket: {
    format: String,
    default: '',
    env: 'GCS_BUCKET'
  },
  expire_times_seconds: {
    format: 'positive-int-array',
    default: [300, 3600, 86400, 604800, 0],
    env: 'EXPIRE_TIMES_SECONDS'
  },
  default_expire_seconds: {
    format: Number,
    default: 86400,
    env: 'DEFAULT_EXPIRE_SECONDS'
  },
  max_expire_seconds: {
    format: Number,
    default: 86400 * 7,
    env: 'MAX_EXPIRE_SECONDS'
  },
  download_counts: {
    format: 'positive-int-array',
    default: [1, 2, 3, 4, 5, 20, 50, 100],
    env: 'DOWNLOAD_COUNTS'
  },
  default_downloads: {
    format: Number,
    default: 1,
    env: 'DEFAULT_DOWNLOADS'
  },
  anon_max_expire_seconds: {
    format: Number,
    default: 86400,
    env: 'ANON_MAX_EXPIRE_SECONDS'
  },
  max_downloads: {
    format: Number,
    default: 100,
    env: 'MAX_DOWNLOADS'
  },
  anon_max_downloads: {
    format: Number,
    default: 20,
    env: 'ANON_MAX_DOWNLOADS'
  },
  max_files_per_archive: {
    format: Number,
    default: 64,
    env: 'MAX_FILES_PER_ARCHIVE'
  },
  max_archives_per_user: {
    format: Number,
    default: 16,
    env: 'MAX_ARCHIVES_PER_USER'
  },
  redis_host: {
    format: String,
    default: 'mock',
    env: 'REDIS_HOST'
  },
  redis_port: {
    format: Number,
    default: 6379,
    env: 'REDIS_PORT'
  },
  redis_user: {
    format: String,
    default: '',
    env: 'REDIS_USER'
  },
  redis_password: {
    format: String,
    default: '',
    env: 'REDIS_PASSWORD'
  },
  redis_db: {
    format: String,
    default: '',
    env: 'REDIS_DB'
  },
  redis_event_expire: {
    format: Boolean,
    default: false,
    env: 'REDIS_EVENT_EXPIRE'
  },
  redis_retry_time: {
    format: Number,
    default: 10000,
    env: 'REDIS_RETRY_TIME'
  },
  redis_retry_delay: {
    format: Number,
    default: 500,
    env: 'REDIS_RETRY_DELAY'
  },
  listen_address: {
    format: 'ipaddress',
    default: '0.0.0.0',
    env: 'IP_ADDRESS'
  },
  listen_port: {
    format: 'port',
    default: 1443,
    arg: 'port',
    env: 'PORT'
  },
  sentry_id: {
    format: String,
    default: '',
    env: 'SENTRY_CLIENT'
  },
  sentry_dsn: {
    format: String,
    default: '',
    env: 'SENTRY_DSN'
  },
  env: {
    format: ['production', 'development', 'test'],
    default: 'development',
    env: 'NODE_ENV'
  },
  max_file_size: {
    format: Number,
    default: 1024 * 1024 * 1024 * 2.5,
    env: 'MAX_FILE_SIZE'
  },
  l10n_dev: {
    format: Boolean,
    default: false,
    env: 'L10N_DEV'
  },
  base_url: {
    format: 'url',
    default: 'https://send.firefox.com',
    env: 'BASE_URL'
  },
  detect_base_url: {
    format: Boolean,
    default: false,
    env: 'DETECT_BASE_URL'
  },
  file_dir: {
    format: 'String',
    default: `${tmpdir()}${path.sep}send-${randomBytes(4).toString('hex')}`,
    env: 'FILE_DIR'
  },
  fxa_required: {
    format: Boolean,
    default: false,
    env: 'FXA_REQUIRED'
  },
  jwe_required: {
    format: Boolean,
    default: true,
    env: 'JWE_REQUIRED'
  },
  fxa_url: {
    format: 'url',
    default: 'https://send-fxa.dev.lcip.org',
    env: 'FXA_URL'
  },
  fxa_client_id: {
    format: String,
    default: '', // disabled
    env: 'FXA_CLIENT_ID'
  },
  fxa_key_scope: {
    format: String,
    default: 'https://identity.mozilla.com/apps/send',
    env: 'FXA_KEY_SCOPE'
  },
  fxa_csp_oauth_url: {
    format: String,
    default: '',
    env: 'FXA_CSP_OAUTH_URL'
  },
  fxa_csp_content_url: {
    format: String,
    default: '',
    env: 'FXA_CSP_CONTENT_URL'
  },
  fxa_csp_profile_url: {
    format: String,
    default: '',
    env: 'FXA_CSP_PROFILE_URL'
  },
  fxa_csp_profileimage_url: {
    format: String,
    default: '',
    env: 'FXA_CSP_PROFILEIMAGE_URL'
  },
  survey_url: {
    format: String,
    default: '',
    env: 'SURVEY_URL'
  },
  ip_db: {
    format: String,
    default: '',
    env: 'IP_DB'
  },
  footer_donate_url: {
    format: String,
    default: '',
    env: 'SEND_FOOTER_DONATE_URL'
  },
  footer_cli_url: {
    format: String,
    default: 'https://github.com/timvisee/ffsend',
    env: 'SEND_FOOTER_CLI_URL'
  },
  footer_dmca_url: {
    format: String,
    default: '',
    env: 'SEND_FOOTER_DMCA_URL'
  },
  footer_source_url: {
    format: String,
    default: 'https://github.com/timvisee/send',
    env: 'SEND_FOOTER_SOURCE_URL'
  },
  ui_color_primary: {
    format: String,
    default: '#0a84ff',
    env: 'UI_COLOR_PRIMARY'
  },
  ui_color_accent: {
    format: String,
    default: '#003eaa',
    env: 'UI_COLOR_ACCENT'
  },
  ui_custom_assets: {
    android_chrome_192px: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_ANDROID_CHROME_192PX'
    },
    android_chrome_512px: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_ANDROID_CHROME_512PX'
    },
    apple_touch_icon: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_APPLE_TOUCH_ICON'
    },
    favicon_16px: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_FAVICON_16PX'
    },
    favicon_32px: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_FAVICON_32PX'
    },
    icon: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_ICON'
    },
    safari_pinned_tab: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_SAFARI_PINNED_TAB'
    },
    facebook: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_FACEBOOK'
    },
    twitter: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_TWITTER'
    },
    wordmark: {
      format: String,
      default: '',
      env: 'UI_CUSTOM_ASSETS_WORDMARK'
    }
  },
  log: {
    app: {
      format: String,
      default: 'FirefoxSend',
      env: 'LOG_APP_NAME'
    },
    level: {
      format: String,
      default: 'info',
      env: 'LOG_LEVEL'
    },
    fmt: {
      format: ['heka', 'pretty'],
      default: 'pretty',
      env: 'LOG_FORMAT'
    }
  },
  access_log: {
    enabled: {
      format: Boolean,
      default: false,
      env: 'ACCESS_LOG_ENABLED'
    },
    fmt: {
      format: ['combined', 'common', 'dev', 'short', 'tiny'],
      default: 'tiny',
      env: 'ACCESS_LOG_FORMAT'
    }
  }
});

// handle configuration files.  you can specify a CSV list of configuration
// files to process, which will be overlayed in order, in the CONFIG_FILES
// environment variable.

let envConfig = path.join(__dirname, `${conf.get('env')}.json`);
envConfig = `${envConfig},/etc/mozilla/send/${conf.get('env')}.json,${process
  .env.CONFIG_FILES || ''}`;
const files = envConfig
  .split(',')
  .map(path => path.trim())
  .filter(path => path.length > 0 && fs.existsSync(path));
conf.loadFile(files);

// Perform validation
conf.validate({ allowed: 'strict' });

const props = conf.getProperties();

const deriveBaseUrl = req => {
  if (!props.detect_base_url) {
    return props.base_url;
  }

  const protocol = req.secure ? 'https://' : 'http://';
  return `${protocol}${req.headers.host}`;
};

module.exports = {
  ...props,
  deriveBaseUrl
};
